### PR TITLE
Remove all non-Fabric-owned version numbers

### PR DESCRIPTION
**Pull request by ohayon**
**TL;DR** _(added by @alloy)_ Twitter has officially taken over maintenance of the Fabric pods, which is good in the long run, but might have some transition issues for you if you were using the unofficial ones before. For a _temporary_ workaround add [this](https://github.com/orta/FabricSpecs) spec repo, but be sure to migrate to the official pods asap.

See [this comment](https://github.com/CocoaPods/Specs/pull/13313#issuecomment-121725141) for more info.

Original text:

---

We want to clean up the history of versions that we (Fabric) did not push ourselves. Some of the version numbers are mismatched and we want to get ahead of some other unforeseen issues, so it would be really awesome if we could only support the versions that we are pushing.

### CODE REVIEW

**Comment by hemaljshah** [comment]
cc @orta - thank you for all the help!

**Comment by keith** [comment]
This is kind of frightening since it will break most existing user's projects.

**Comment by ohwutup** [comment]
@keith - Now that we have taken ownership over the pods we would like to cleanup the unsupported versions. We want to allow all developers to use Fabric and CocoaPods from their official sources rather than a third party who may be tampering with the frameworks before putting them under our trusted name.

**Comment by keith** [comment]
Actually removing the 1.3.x versions actually broke our project.

**Comment by hemaljshah** [comment]
@keith were you pointing to a specific version? Version 1.3.2 isn't even a supported version on our side, we're currently at 1.2.8 :)

**Comment by keith** [comment]
Previously the most recent spec was 1.3.2. Those were removed here: https://github.com/CocoaPods/Specs/commit/562913080d02794958de1eeca6632b82db0c79be

**Comment by hemaljshah** [comment]
Yup, that was previously posted by https://zergtmn.github.io/ before we had ownership of this pod. Part of having ownership is the clean up from unsupported versioning and level setting to what we are shipping to our customers manually in the Mac app.

We want all of our customers to be on the same versioning to avoid confusion & ensure they are using a pod directly created by us without any potential chance for tampering.

**Comment by keith** [comment]
Also it seems as though Crashlytics implicitly depends on Fabric but that is not in the spec? For example in a fresh project just including Crashlytics does not compile. Is this intended?

**Comment by orta** [comment]
Yeah, it's a bit of a transition, but better for everyone in the long run alas. 
Re: implicit dependencies, looks like the example on the site just recommends [including them all](http://docs.fabric.io/ios/fabric/cocoapods.html).

**Comment by ohwutup** [comment]
Yes, we realize this is a bit awkward but for now this is intended to get around the transitive dependencies issue. We will likely add these dependencies in once 0.38.0 is released.

**Comment by orta** [comment]
:+1:

**Comment by orta** [comment]
To people reading this thread later, you may have previously had a Podfile like 
```
pod "Fabric/Crashlytics"
```
To migrate, change to:

```
pod ‘Fabric’
pod ‘Crashlytics’
pod ‘Digits’
pod ‘TwitterKit’
pod ‘MoPub’
```

For more details see [http://docs.fabric.io/ios/fabric/cocoapods.html](docs.fabric.io)

**Comment by ohwutup** [comment]
Thanks @orta ! Also, our Mac App will walk you through the appropriate setup for podfiles during onboarding.

**Comment by orta** [comment]
Ace

**Comment by ohwutup** [comment]
@orta - one more thing! Is there anything we can do on our end to fix up the versions and results on cocoapods.org? Versions look wacky and some pods we just removed are still showing up. Happy to help out however we can!

**Comment by bpoplauschi** [comment]
Guys, I have been maintaining CrashlyticsFramework as a pod until now. I have [deprecated it](https://guides.cocoapods.org/syntax/podspec.html#deprecated_in_favor_of) in favor of you official pods. Final step is updating all the pod specs here with the deprecated flag (https://github.com/CocoaPods/Specs/pull/13314). 
Great job adding support for CocoaPods.

**Comment by hemaljshah** [comment]
Thank you @bpoplauschi. Appreciate the support, we're excited to continue to meet developers where they are comfortable and support the tools they love to use!

**Comment by keith** [comment]
Can we go ahead and add Fabric as a Dependecy of Crashlytics? This seems like a configuration error.

**Comment by orta** [comment]
Search takes 1ish minutes to update it's db.

**Comment by ohwutup** [comment]
@keith - Totally understand this is weird but it is by design for the time being because of the transitive dependencies issues in 0.37.2. We will definitely reevaluate once 0.38.0 ships.

**Comment by hemaljshah** [comment]
@orta Awesome, still looks it's populating multiple lines of results for the OS X pod and the Fabric pod has old README instructions on the page.

**Comment by keith** [comment]
Might want to keep an eye out for the fallout of breaking projects https://mobile.twitter.com/kleinlieu/status/621398686635929600

**Comment by hemaljshah** [comment]
@keith Absolutely. And our support teams are working on documentation to help users properly migrate from unofficial pods to the officially supported configuration.

**Comment by supermarin** [comment]
What just happened is completely unacceptable.
You broke all the projects using Crashlytics.

**Comment by keith** [comment]
People relying on Crashlytics should probably just include:

```
pod 'Fabric'
pod 'Crashlytics'
```

The others are unrelated and unnecessary for using Craslytics.

**Comment by jhersh** [comment]
"Not to pile on, but..." he began, piling on, "This was handled poorly."

I think visitors in the near future will be keen on this link (the docs URL several comments above is incorrect): http://docs.fabric.io/ios/fabric/cocoapods.html

And for the immediate-term need to get your project building again, are we quite certain that anyone using a single Crashlytics pod previously will now need to include *two* pods in their project, both `Fabric` and `Crashlytics`?

**Comment by mergesort** [comment]
@jhersh Yes. `Crashlytics` has references to macros in `FABAttributes.h`, which is only in Fabric. Once I added Fabric, everything worked swell.

**Comment by sberrevoets** [comment]
@hemaljshah 
> And our support teams are working on documentation to help users properly migrate from unofficial pods to the officially supported configuration.

That should have been done before making this change.

**Comment by getaaron** [comment]
@orta Your example uses curly quotes (`pod ‘Fabric’`) which won't work. Can you please edit your comment to replace them with straight quotes (`pod 'Fabric'`)?

**Comment by getaaron** [comment]
@ohwutup The docs that @jhersh [linked to](http://docs.fabric.io/ios/fabric/cocoapods.html) also include curly quotes. You should replace them with straight quotes.

**Comment by ohwutup** [comment]
Thanks @getaaron ! We are fixing it by just putting the call in quotes to make sure that paths with spaces are respected.

**Comment by dcow** [comment]
This is absolutely and 100% unacceptable. There is no way to explain your way out of this blunder. You just broke all old builds that use cocoapods. I can't go back to release 1.2.x, branch, and cherry-pick a hotfix without also pulling an update for this. No heads up, no warning that this would be deprecated. I can't build any old sha. It's all broken. This is what happens when people don't understand referential transparency. Fuck state.

**Comment by segiddins** [comment]
@dcow this is your one warning. That sort of language is not tolerated here.

**Comment by dcow** [comment]
@segiddins I'm sorry if I offended anyone. Are there community guidelines you can point me at regarding explicit language; my bad.

**Comment by segiddins** [comment]
Our code of conduct is at https://github.com/CocoaPods/CocoaPods/blob/master/CODE_OF_CONDUCT.md.

**Comment by dcow** [comment]
Looks like language isn't mentioned.. @segiddins here you go: https://github.com/dcow/CocoaPods/pull/1 (=

**Comment by dcow** [comment]
I've opened https://github.com/CocoaPods/CocoaPods/issues/3827 to facilitate discussion around how we can avoid issues like this in the future.

**Comment by getaaron** [comment]
Should we consider putting the unofficial pods back in and deprecating them as described in https://github.com/CocoaPods/Specs/pull/12199?

**Comment by dcow** [comment]
I support that.

**Comment by DenHeadless** [comment]
I see no reason why podspecs were deleted instead of deprecating them. We too have an internal SDK, that integrates crashlytics, and applications, that pull that SDK. As of now, there's no way of compiling projects unless updating all of them to new version of Fabric.

Of course, this would need to be done either way, but transition period or heads-up would be great before  doing such destructive action as deleting old podspecs.

I think CocoaPods authors should be aware of dependency system fragility, as compared to RubyGems, for example. This kind of stuff encourages users to think, that keeping Pods directory under source control is absolutely neccessary, because some day one of your pods will break, and you will wonder "Why did i consider Pods directory a build artifact in the first place?"

**Comment by orta** [comment]
Anyone with backwards compatibility problems can use the steps in [this repo](https://github.com/orta/FabricSpecs) to get the exact same behaviour from CocoaPods. I've moved the deprecated Specs into a separate repo that can be prioritised as being more important when resolving for a Pod.

I understand it's frustrating, and we _always_ value system stability with the trade-off of saying no to almost every one of these requests. These circumstances are unique though. In an ideal world the unofficial pods from before would have called themselves that in their name. It's not unreasonable to consider that the Fabric team wouldn't want anyone resolving to an older build on CocoaPods and then having people using binaries/scripts that are totally out of their control. 

With respect to temporarily bringing back and deprecating the older Pods for a limited time, that choice is on the Fabric team. In the mean time my [Specs repo](https://github.com/orta/FabricSpecs) can be used to continue as it was yesterday.

**Comment by valeriomazzeo** [comment]
... and this is why you should check in the Pods directory my friends ....

**Comment by dcow** [comment]
So this appears to have been a double blunder. I should have directed some of my frustration at the Fabric team instead (; Here's a fix for the Crashlytics pod spec: https://github.com/CocoaPods/Specs/pull/13315

In short, the Fabric team forgot to enumerate the public headers so during the install CocoaPods doesn't know to link the headers from the framework into the public headers directory.

I don't know if I should contact the Fabric team to get them to update, but fixing this would probably help others who may be using Crashlytics from within a static library pod. I'm not sure why it works in the normal case. Perhaps Cocoapods automatically detects headers in top level `vendored_frameworks`?

Now, just gotta fix Fabric's obsession with http *no* s. Heads up Crashlytics users, you're downloading precompiled code over insecure connections whenever you install or update Fabric tools.

**Comment by alloy** [comment]
@ohwutup I’ve updated your comment with a link to @orta’s workaround spec repo at the top, so that people that just need to get their current builds fixed get the right answers asap. Let me know if either of you feel it needs changing.

**Comment by dcow** [comment]
@orta @alloy thanks for cleaning this up promptly and finding some middle ground to work with. As you say, the real culprit is whoever ordered the old repos be taken down. But damn I wish someone somewhere had asked how many users are still pulling in the old versions before removing them. This isn't the first time we've lost a day to Crashlytics' build eccentrics. Cheers to catching up tomorrow 🍻

**Comment by crsantos** [comment]
@orta Any ideas for who projects that use cocoapods 0.33.1 that doesn't allow setting "source"?

**Comment by alloy** [comment]
@crsantos Vendor the removed specs in your code repo and use them directly from there https://guides.cocoapods.org/syntax/podfile.html#podspec

**Comment by orta** [comment]
@crsantos it's probably best that you you use [Podspec](https://guides.cocoapods.org/syntax/podfile.html#podspec) directly from my repo, something like: 
```
pod "Fabric" , :podspec => "https://raw.githubusercontent.com/orta/FabricSpecs/master/Specs/Fabric/1.2/Fabric.podspec.json"
```

**Comment by crsantos** [comment]
@orta Thank you so much!

**Comment by ohwutup** [comment]
@orta, @alloy, @dcow - thank you guys for your help and pointing out some fixes. Here is a PR that fixes the HTTPS issue. https://github.com/CocoaPods/Specs/pull/13316

**Comment by billinghamj** [comment]
This is completely unacceptable. Whether or not the Fabric team wanted them gone, the CocoaPods team should not have allowed it.

This has broken our build and caused a load of hassle & wasted time.

**Comment by billinghamj** [comment]
Also cocoapods.org is still showing 1.3.2 as the latest version - https://cocoapods.org/pods/Fabric

**Comment by jafara** [comment]
LOL. What a mess. You guys just broke every single release we've done in the past that depends on these podspecs. This mean no more hermetic releases and we need to update our Podfile with some other new version that doesn't guarantee to match the sources we've used when we built the release.

**Comment by erickreutz** [comment]
@orta your spec repo doesn't contain the `~> 1.3` specs that use to be here.

**Comment by keith** [comment]
@erickreutz you'll have to "downgrade" to the newest available version. Apparently the unofficial version pushed to CocoaPods diverged from Twitter's internal versions.

**Comment by drn** [comment]
Very frustrating. Builds stopped working for no reason on our CI server due to this issue. Wasted morning... Hope this gets resolved soon.
@orta - Really appreciate the temporary workaround.

**Comment by shepting** [comment]
I believe you should be able to just update your `Podfile` to contain:

    pod 'Fabric'
    pod 'Crashlytics'

**Comment by keith** [comment]
Unless this is reverted this workaround is not temporary.

**Comment by dcow** [comment]
@shepting @keith it's really not that simple. The mentality that this change is acceptable is pretty presumptuous because you assume that you know the details of everyones' build. 

* What if I depend on a specific version of the pod's binary? 
* Maybe I'm using undocumented features *oh gosh* and it is cumbersome or impossible to update.
* What if I no longer use the pod in my build but need to go back and release a hotfix?
* What if there is a security audit process required for any 3rd party dependency updates?
* What if the license changes and I cannot update to a newer version for legal reasons?
* I'm sure there are more.

For those who aren't single-machine devs writing sexy apps using bleeding edge software, or anyone who has large build systems setup with many different release versions to manage, these scenarios are all too common. For enterprise focused teams who's business literally depends on meeting contractual deadlines with customers, at the wrong time an incident like this is *extremely* painful, and could even have financial and legal ramifications.

I know this probably seems like a big lazy developer overreaction, but please take these comments seriously. Maybe this is fine in the average use case, but for a nontrivial subset of teams and developers, this is entirely not okay. I now have to discuss with my team if this "third party community project" CocoaPods is mature enough to rely on in business critical build pathways.

**Comment by shepting** [comment]
@dcow Definitely.

I think the big change was the move from unsupported to supported.

If you wanted a specific version of Crashlytics before, the only supported way to get the framework was the Fabric Mac app. Now you can use CocoaPods to get the direct version you desire.

If you received a bad version from the Mac app you didn't have any way to go back, now you can use CocoaPods.

I understand that you want to be able to go back to old builds, but the recommended way to ensure you could re-create an old setup was ***always*** to check-in your Pods. I know it doesn't make things easier *right now*, but it's a good practice for any team that needs to have reproducible builds (which I would argue is *every* team).

This *was* certainly a big change for developers relying on unsupported and unofficial Podspecs. It *was* unfortunate that Fabric didn't release a supported Podspec before this. But long-term this will give you all the things that you want in that list.

Looking at the big picture I think this is a really positive move, that one of the last hold-outs from supplying a CocoaPod has finally decided to give developers a choice to use their preferred dependency manager.

**Comment by keith** [comment]
Another thing to note here, if you were relying on the binary path to submit your app to crashlytics the path of that has changed as well.


From:

```
Pods/Fabric/Crashlytics.framework/submit
```

To:

```
Pods/Crashlytics/Crashlytics.framework/submit
```

**Comment by neonichu** [comment]
Let me just drop <https://github.com/xing/XNGPodsSynchronizer> here, which might help you to ensure stability of historic builds. Also don't forget that there is generally no guarantee that a Pod's sources are available at any point in time, because they are hosted by each individual author. Relying on the master repo and the availability of historic sources has never been a good idea.

**Comment by alloy** [comment]
This ticket has now had inputs from all possible points of view, so for now I’m going to lock this issue.

Could the migration have gone differently? Possibly, yeah. We’re _definitely_ evaluating everything to see if things can be done smoother in the future, so thanks all for taking the time to leave feedback.
